File: src/web-ui/src/infrastructure/api/service-api/GitAPI.ts
 

import { api } from './ApiClient';
import { createTauriCommandError } from '../errors/TauriCommandError';
import { createLogger } from '@/shared/utils/logger';

const log = createLogger('GitAPI');

export interface GitRepository {
  path: string;
  name: string;
  current_branch: string;
  is_bare: boolean;
  has_changes: boolean;
  remotes: string[];
  
  branch?: string;
  remote?: string;
  lastCommit?: string;
}


export interface GitFileStatusDetail {
   
  path: string;
   
  status: string;
   
  index_status?: string;
   
  workdir_status?: string;
}

export interface GitStatus {
  staged: GitFileStatusDetail[];
  unstaged: GitFileStatusDetail[];
  untracked: string[];
  conflicts: string[];
  current_branch: string;
  ahead: number;
  behind: number;
}

export interface GitCommit {
  hash: string;
  message: string;
  author: string;
  date: string;
  files?: string[];
}

export interface GitBranch {
  name: string;
  current: boolean;
  remote: boolean;
  lastCommit?: string;
  ahead?: number;
  behind?: number;
}

export interface GitOperationResult {
  success: boolean;
  message?: string;
  error?: string;
}

export interface GitAddParams {
  files: string[];
  all?: boolean;
}

export interface GitCommitParams {
  message: string;
  amend?: boolean;
  signoff?: boolean;
}

export interface GitPushParams {
  remote?: string;
  branch?: string;
  force?: boolean;
  setUpstream?: boolean;
}

export interface GitPullParams {
  remote?: string;
  branch?: string;
  rebase?: boolean;
}

export interface GitDiffParams {
  source?: string;
  target?: string;
  files?: string[];
  stat?: boolean;
  filePath?: string;
  staged?: boolean;
  commit?: string;
}

export interface GitChangedFilesParams {
  source?: string;
  target?: string;
  staged?: boolean;
}

export type GitChangedFileStatus =
  | 'added'
  | 'modified'
  | 'deleted'
  | 'renamed'
  | 'copied'
  | 'unknown';

export interface GitChangedFile {
  path: string;
  old_path?: string;
  status: GitChangedFileStatus;
}

export interface GitLogParams {
  maxCount?: number;
  since?: string;
  until?: string;
  author?: string;
}

export interface GitOperationParams {
  repositoryPath: string;
  command: string;
  args?: string[];
}


export interface GitGraphRef {
  name: string;
  refType: 'branch' | 'remote' | 'tag';
  isCurrent: boolean;
  isHead: boolean;
}

export interface GitGraphNode {
  hash: string;
  message: string;
  fullMessage: string;
  authorName: string;
  authorEmail: string;
  timestamp: number;
  parents: string[];
  children: string[];
  refs: GitGraphRef[];
  lane: number;
  forkingLanes: number[];
  mergingLanes: number[];
  passingLanes: number[];
}

export interface GitGraph {
  nodes: GitGraphNode[];
  maxLane: number;
  currentBranch?: string;
}


export interface GitWorktreeInfo {
   
  path: string;
   
  branch: string | null;
  /** HEAD commit hash */
  head: string;
   
  isMain: boolean;
   
  isLocked: boolean;
   
  isPrunable: boolean;
}

export class GitAPI {
   
  async isGitRepository(repositoryPath: string): Promise<boolean> {
    try {
      return await api.invoke('git_is_repository', { 
        request: { repositoryPath } 
      });
    } catch (error) {
      throw createTauriCommandError('git_is_repository', error, { repositoryPath });
    }
  }

   
  async getRepository(repositoryPath: string): Promise<GitRepository> {
    try {
      return await api.invoke('git_get_repository', { 
        request: { repositoryPath } 
      });
    } catch (error) {
      throw createTauriCommandError('git_get_repository', error, { repositoryPath });
    }
  }

   
  async getStatus(repositoryPath: string): Promise<GitStatus> {
    try {
      return await api.invoke('git_get_status', { 
        request: { repositoryPath } 
      });
    } catch (error) {
      throw createTauriCommandError('git_get_status', error, { repositoryPath });
    }
  }

   
  async getBranches(repositoryPath: string, includeRemote: boolean = false): Promise<GitBranch[]> {
    try {
      return await api.invoke('git_get_branches', { 
        request: { repositoryPath, includeRemote } 
      });
    } catch (error) {
      throw createTauriCommandError('git_get_branches', error, { repositoryPath, includeRemote });
    }
  }

   
  async getEnhancedBranches(repositoryPath: string, includeRemote: boolean = false): Promise<GitBranch[]> {
    try {
      return await api.invoke('git_get_enhanced_branches', { 
        request: { repositoryPath, includeRemote } 
      });
    } catch (error) {
      throw createTauriCommandError('git_get_enhanced_branches', error, { repositoryPath, includeRemote });
    }
  }

   
  async getCommits(repositoryPath: string, params: GitLogParams = {}): Promise<GitCommit[]> {
    try {
      return await api.invoke('git_get_commits', { 
        request: { repositoryPath, params } 
      });
    } catch (error) {
      throw createTauriCommandError('git_get_commits', error, { repositoryPath, params });
    }
  }

   
  async addFiles(repositoryPath: string, params: GitAddParams): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_add_files', { 
        request: { repositoryPath, params } 
      });
    } catch (error) {
      throw createTauriCommandError('git_add_files', error, { repositoryPath, params });
    }
  }

   
  async commit(repositoryPath: string, params: GitCommitParams): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_commit', { 
        request: { repositoryPath, params } 
      });
    } catch (error) {
      throw createTauriCommandError('git_commit', error, { repositoryPath, params });
    }
  }

   
  async push(repositoryPath: string, params: GitPushParams = {}): Promise<GitOperationResult> {
    try {
      
      const backendParams = {
        remote: params.remote,
        branch: params.branch,
        force: params.force,
        set_upstream: params.setUpstream
      };
      
      return await api.invoke('git_push', { 
        request: { repositoryPath, params: backendParams } 
      });
    } catch (error) {
      throw createTauriCommandError('git_push', error, { repositoryPath, params });
    }
  }

   
  async pull(repositoryPath: string, params: GitPullParams = {}): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_pull', { 
        request: { repositoryPath, params } 
      });
    } catch (error) {
      throw createTauriCommandError('git_pull', error, { repositoryPath, params });
    }
  }

   
  async checkoutBranch(repositoryPath: string, branchName: string): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_checkout_branch', { 
        request: { repositoryPath, branchName } 
      });
    } catch (error) {
      throw createTauriCommandError('git_checkout_branch', error, { repositoryPath, branchName });
    }
  }

   
  async createBranch(repositoryPath: string, branchName: string, startPoint?: string): Promise<GitOperationResult> {
    try {
      
      const effectiveStartPoint = startPoint && startPoint.trim() ? startPoint : undefined;
      return await api.invoke('git_create_branch', { 
        request: { repositoryPath, branchName, startPoint: effectiveStartPoint } 
      });
    } catch (error) {
      throw createTauriCommandError('git_create_branch', error, { repositoryPath, branchName, startPoint });
    }
  }

   
  async deleteBranch(repositoryPath: string, branchName: string, force: boolean = false): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_delete_branch', { 
        request: { repositoryPath, branchName, force } 
      });
    } catch (error) {
      throw createTauriCommandError('git_delete_branch', error, { repositoryPath, branchName, force });
    }
  }

   
  async resetToCommit(repositoryPath: string, commitHash: string, mode: 'soft' | 'mixed' | 'hard' = 'mixed'): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_reset_to_commit', { 
        request: { repositoryPath, commitHash, mode } 
      });
    } catch (error) {
      throw createTauriCommandError('git_reset_to_commit', error, { repositoryPath, commitHash, mode });
    }
  }

   
  async getDiff(repositoryPath: string, params: GitDiffParams): Promise<string> {
    try {
      return await api.invoke('git_get_diff', { 
        request: { repositoryPath, params } 
      });
    } catch (error) {
      throw createTauriCommandError('git_get_diff', error, { repositoryPath, params });
    }
  }

   
  async getChangedFiles(repositoryPath: string, params: GitChangedFilesParams): Promise<GitChangedFile[]> {
    try {
      return await api.invoke('git_get_changed_files', {
        request: { repositoryPath, params }
      });
    } catch (error) {
      throw createTauriCommandError('git_get_changed_files', error, { repositoryPath, params });
    }
  }


  async resetFiles(repositoryPath: string, files: string[], staged: boolean = false): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_reset_files', { 
        request: { repositoryPath, files, staged } 
      });
    } catch (error) {
      throw createTauriCommandError('git_reset_files', error, { repositoryPath, files, staged });
    }
  }

   
  async getFileContent(repositoryPath: string, filePath: string, commit?: string): Promise<string> {
    try {
      return await api.invoke('git_get_file_content', { 
        request: { repositoryPath, filePath, commit } 
      });
    } catch (error) {
      throw createTauriCommandError('git_get_file_content', error, { repositoryPath, filePath, commit });
    }
  }
   
  async getGraph(repositoryPath: string, maxCount?: number, branchName?: string): Promise<GitGraph> {
    try {
      const result = await api.invoke<GitGraph>('git_get_graph', { 
        repositoryPath,
        maxCount: maxCount || null,
        branchName: branchName || null
      });
      return result;
    } catch (error) {
      log.error('Failed to get git graph', { repositoryPath, maxCount, branchName, error });
      throw createTauriCommandError('git_get_graph', error, { repositoryPath, maxCount, branchName });
    }
  }

   
  async cherryPick(repositoryPath: string, commitHash: string, noCommit: boolean = false): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_cherry_pick', { 
        request: { repositoryPath, commitHash, noCommit } 
      });
    } catch (error) {
      throw createTauriCommandError('git_cherry_pick', error, { repositoryPath, commitHash, noCommit });
    }
  }

   
  async cherryPickAbort(repositoryPath: string): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_cherry_pick_abort', { 
        request: { repositoryPath } 
      });
    } catch (error) {
      throw createTauriCommandError('git_cherry_pick_abort', error, { repositoryPath });
    }
  }

   
  async cherryPickContinue(repositoryPath: string): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_cherry_pick_continue', { 
        request: { repositoryPath } 
      });
    } catch (error) {
      throw createTauriCommandError('git_cherry_pick_continue', error, { repositoryPath });
    }
  }

  

   
  async listWorktrees(repositoryPath: string): Promise<GitWorktreeInfo[]> {
    try {
      return await api.invoke('git_list_worktrees', { 
        request: { repositoryPath } 
      });
    } catch (error) {
      throw createTauriCommandError('git_list_worktrees', error, { repositoryPath });
    }
  }

   
  async addWorktree(repositoryPath: string, branch: string, createBranch: boolean = false): Promise<GitWorktreeInfo> {
    try {
      return await api.invoke('git_add_worktree', { 
        request: { repositoryPath, branch, createBranch } 
      });
    } catch (error) {
      throw createTauriCommandError('git_add_worktree', error, { repositoryPath, branch, createBranch });
    }
  }

   
  async removeWorktree(repositoryPath: string, worktreePath: string, force: boolean = false): Promise<GitOperationResult> {
    try {
      return await api.invoke('git_remove_worktree', { 
        request: { repositoryPath, worktreePath, force } 
      });
    } catch (error) {
      throw createTauriCommandError('git_remove_worktree', error, { repositoryPath, worktreePath, force });
    }
  }
}


export const gitAPI = new GitAPI();

File: src/apps/desktop/src/api/git_api.rs
//! Git API

use crate::api::app_state::AppState;
use bitfun_core::infrastructure::storage::StorageOptions;
use bitfun_core::service::git::{
    GitAddParams, GitChangedFile, GitChangedFilesParams, GitCommitParams, GitDiffParams,
    GitLogParams, GitPullParams, GitPushParams, GitService,
};
use bitfun_core::service::git::{
    GitBranch, GitCommit, GitOperationResult, GitRepository, GitStatus,
};
use log::{error, info};
use serde::{Deserialize, Serialize};
use tauri::State;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitRepositoryRequest {
    pub repository_path: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitBranchesRequest {
    pub repository_path: String,
    pub include_remote: Option<bool>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCommitsRequest {
    pub repository_path: String,
    pub params: Option<GitLogParams>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitAddFilesRequest {
    pub repository_path: String,
    pub params: GitAddParams,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCommitRequest {
    pub repository_path: String,
    pub params: GitCommitParams,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitPushRequest {
    pub repository_path: String,
    pub params: GitPushParams,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitPullRequest {
    pub repository_path: String,
    pub params: GitPullParams,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCheckoutBranchRequest {
    pub repository_path: String,
    pub branch_name: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCreateBranchRequest {
    pub repository_path: String,
    pub branch_name: String,
    pub start_point: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitDeleteBranchRequest {
    pub repository_path: String,
    pub branch_name: String,
    pub force: Option<bool>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitDiffRequest {
    pub repository_path: String,
    pub params: GitDiffParams,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitChangedFilesRequest {
    pub repository_path: String,
    pub params: GitChangedFilesParams,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitResetFilesRequest {
    pub repository_path: String,
    pub files: Vec<String>,
    pub staged: Option<bool>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitResetToCommitRequest {
    pub repository_path: String,
    pub commit_hash: String,
    pub mode: String, // "soft", "mixed", or "hard"
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitGetFileContentRequest {
    pub repository_path: String,
    pub file_path: String,
    pub commit: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCherryPickRequest {
    pub repository_path: String,
    pub commit_hash: String,
    pub no_commit: Option<bool>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitAddWorktreeRequest {
    pub repository_path: String,
    pub branch: String,
    pub create_branch: Option<bool>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitRemoveWorktreeRequest {
    pub repository_path: String,
    pub worktree_path: String,
    pub force: Option<bool>,
}

#[tauri::command]
pub async fn git_is_repository(
    _state: State<'_, AppState>,
    request: GitRepositoryRequest,
) -> Result<bool, String> {
    GitService::is_repository(&request.repository_path)
        .await
        .map_err(|e| {
            error!(
                "Failed to check Git repository: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to check Git repository: {}", e)
        })
}

#[tauri::command]
pub async fn git_get_repository(
    _state: State<'_, AppState>,
    request: GitRepositoryRequest,
) -> Result<GitRepository, String> {
    GitService::get_repository(&request.repository_path)
        .await
        .map_err(|e| {
            error!(
                "Failed to get Git repository info: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to get Git repository info: {}", e)
        })
}

#[tauri::command]
pub async fn git_get_status(
    _state: State<'_, AppState>,
    request: GitRepositoryRequest,
) -> Result<GitStatus, String> {
    GitService::get_status(&request.repository_path)
        .await
        .map_err(|e| {
            error!(
                "Failed to get Git status: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to get Git status: {}", e)
        })
}

#[tauri::command]
pub async fn git_get_branches(
    _state: State<'_, AppState>,
    request: GitBranchesRequest,
) -> Result<Vec<GitBranch>, String> {
    let include_remote = request.include_remote.unwrap_or(false);
    GitService::get_branches(&request.repository_path, include_remote)
        .await
        .map_err(|e| {
            error!(
                "Failed to get Git branches: path={}, include_remote={}, error={}",
                request.repository_path, include_remote, e
            );
            format!("Failed to get Git branches: {}", e)
        })
}

#[tauri::command]
pub async fn git_get_enhanced_branches(
    _state: State<'_, AppState>,
    request: GitBranchesRequest,
) -> Result<Vec<GitBranch>, String> {
    let include_remote = request.include_remote.unwrap_or(false);
    GitService::get_enhanced_branches(&request.repository_path, include_remote)
        .await
        .map_err(|e| {
            error!(
                "Failed to get enhanced Git branches: path={}, include_remote={}, error={}",
                request.repository_path, include_remote, e
            );
            format!("Failed to get enhanced Git branches: {}", e)
        })
}

#[tauri::command]
pub async fn git_get_commits(
    _state: State<'_, AppState>,
    request: GitCommitsRequest,
) -> Result<Vec<GitCommit>, String> {
    let params = request.params.unwrap_or_default();
    GitService::get_commits(&request.repository_path, params)
        .await
        .map_err(|e| {
            error!(
                "Failed to get Git commits: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to get Git commits: {}", e)
        })
}

#[tauri::command]
pub async fn git_add_files(
    _state: State<'_, AppState>,
    request: GitAddFilesRequest,
) -> Result<GitOperationResult, String> {
    GitService::add_files(&request.repository_path, request.params)
        .await
        .map_err(|e| {
            error!(
                "Failed to add files: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to add files: {}", e)
        })
}

#[tauri::command]
pub async fn git_commit(
    _state: State<'_, AppState>,
    request: GitCommitRequest,
) -> Result<GitOperationResult, String> {
    GitService::commit(&request.repository_path, request.params)
        .await
        .map_err(|e| {
            error!(
                "Failed to commit: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to commit: {}", e)
        })
}

#[tauri::command]
pub async fn git_push(
    _state: State<'_, AppState>,
    request: GitPushRequest,
) -> Result<GitOperationResult, String> {
    GitService::push(&request.repository_path, request.params)
        .await
        .map_err(|e| {
            error!(
                "Failed to push: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to push: {}", e)
        })
}

#[tauri::command]
pub async fn git_pull(
    _state: State<'_, AppState>,
    request: GitPullRequest,
) -> Result<GitOperationResult, String> {
    GitService::pull(&request.repository_path, request.params)
        .await
        .map_err(|e| {
            error!(
                "Failed to pull: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to pull: {}", e)
        })
}

#[tauri::command]
pub async fn git_checkout_branch(
    _state: State<'_, AppState>,
    request: GitCheckoutBranchRequest,
) -> Result<GitOperationResult, String> {
    GitService::checkout_branch(&request.repository_path, &request.branch_name)
        .await
        .map_err(|e| {
            error!(
                "Failed to checkout branch: path={}, branch={}, error={}",
                request.repository_path, request.branch_name, e
            );
            format!("Failed to checkout branch: {}", e)
        })
}

#[tauri::command]
pub async fn git_create_branch(
    _state: State<'_, AppState>,
    request: GitCreateBranchRequest,
) -> Result<GitOperationResult, String> {
    GitService::create_branch(
        &request.repository_path,
        &request.branch_name,
        request.start_point.as_deref(),
    )
    .await
    .map_err(|e| {
        error!(
            "Failed to create branch: path={}, branch={}, error={}",
            request.repository_path, request.branch_name, e
        );
        format!("Failed to create branch: {}", e)
    })
}

#[tauri::command]
pub async fn git_delete_branch(
    _state: State<'_, AppState>,
    request: GitDeleteBranchRequest,
) -> Result<GitOperationResult, String> {
    let force = request.force.unwrap_or(false);
    GitService::delete_branch(&request.repository_path, &request.branch_name, force)
        .await
        .map_err(|e| {
            error!(
                "Failed to delete branch: path={}, branch={}, force={}, error={}",
                request.repository_path, request.branch_name, force, e
            );
            format!("Failed to delete branch: {}", e)
        })
}

#[tauri::command]
pub async fn git_get_diff(
    _state: State<'_, AppState>,
    request: GitDiffRequest,
) -> Result<String, String> {
    GitService::get_diff(&request.repository_path, &request.params)
        .await
        .map_err(|e| {
            error!(
                "Failed to get Git diff: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to get Git diff: {}", e)
        })
}

#[tauri::command]
pub async fn git_get_changed_files(
    _state: State<'_, AppState>,
    request: GitChangedFilesRequest,
) -> Result<Vec<GitChangedFile>, String> {
    info!(
        "Getting changed Git files for repository: {}",
        request.repository_path
    );

    GitService::get_changed_files(&request.repository_path, &request.params)
        .await
        .map_err(|e| {
            error!("Failed to get changed Git files: {}", e);
            e.to_string()
        })
}

#[tauri::command]
pub async fn git_reset_files(
    _state: State<'_, AppState>,
    request: GitResetFilesRequest,
) -> Result<GitOperationResult, String> {
    let staged = request.staged.unwrap_or(false);

    info!(
        "Resetting files in '{}' (staged: {}): {:?}",
        request.repository_path, staged, request.files
    );

    GitService::reset_files(&request.repository_path, &request.files, staged)
        .await
        .map(|output| GitOperationResult {
            success: true,
            data: None,
            error: None,
            output: Some(output),
            duration: None,
        })
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn git_get_file_content(
    _state: State<'_, AppState>,
    request: GitGetFileContentRequest,
) -> Result<String, String> {
    info!(
        "Getting file content for '{}' at commit '{:?}' in repo '{}'",
        request.file_path, request.commit, request.repository_path
    );

    let content = GitService::get_file_content(
        &request.repository_path,
        &request.file_path,
        request.commit.as_deref(),
    )
    .await
    .map_err(|e| e.to_string())?;

    Ok(content)
}

#[tauri::command]
pub async fn git_reset_to_commit(
    _state: State<'_, AppState>,
    request: GitResetToCommitRequest,
) -> Result<GitOperationResult, String> {
    info!(
        "Resetting to commit '{}' with mode '{}' in repo '{}'",
        request.commit_hash, request.mode, request.repository_path
    );

    GitService::reset_to_commit(
        &request.repository_path,
        &request.commit_hash,
        &request.mode,
    )
    .await
    .map_err(|e| {
        error!(
            "Failed to reset to commit: path={}, commit={}, mode={}, error={}",
            request.repository_path, request.commit_hash, request.mode, e
        );
        format!("Failed to reset: {}", e)
    })
}

#[tauri::command]
pub async fn git_get_graph(
    _state: State<'_, AppState>,
    repository_path: String,
    max_count: Option<usize>,
    branch_name: Option<String>,
) -> Result<bitfun_core::service::git::GitGraph, String> {
    info!(
        "Getting git graph: repository_path={}, max_count={:?}, branch_name={:?}",
        repository_path, max_count, branch_name
    );

    GitService::get_git_graph_for_branch(&repository_path, max_count, branch_name.as_deref())
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn git_cherry_pick(
    _state: State<'_, AppState>,
    request: GitCherryPickRequest,
) -> Result<GitOperationResult, String> {
    let no_commit = request.no_commit.unwrap_or(false);

    info!(
        "Cherry-picking commit '{}' in repo '{}' (no_commit: {})",
        request.commit_hash, request.repository_path, no_commit
    );

    GitService::cherry_pick(&request.repository_path, &request.commit_hash, no_commit)
        .await
        .map_err(|e| {
            error!(
                "Failed to cherry-pick: path={}, commit={}, no_commit={}, error={}",
                request.repository_path, request.commit_hash, no_commit, e
            );
            format!("Failed to cherry-pick: {}", e)
        })
}

#[tauri::command]
pub async fn git_cherry_pick_abort(
    _state: State<'_, AppState>,
    request: GitRepositoryRequest,
) -> Result<GitOperationResult, String> {
    info!("Aborting cherry-pick in repo '{}'", request.repository_path);

    GitService::cherry_pick_abort(&request.repository_path)
        .await
        .map_err(|e| {
            error!(
                "Failed to abort cherry-pick: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to abort cherry-pick: {}", e)
        })
}

#[tauri::command]
pub async fn git_cherry_pick_continue(
    _state: State<'_, AppState>,
    request: GitRepositoryRequest,
) -> Result<GitOperationResult, String> {
    info!(
        "Continuing cherry-pick in repo '{}'",
        request.repository_path
    );

    GitService::cherry_pick_continue(&request.repository_path)
        .await
        .map_err(|e| {
            error!(
                "Failed to continue cherry-pick: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to continue cherry-pick: {}", e)
        })
}

#[tauri::command]
pub async fn git_list_worktrees(
    _state: State<'_, AppState>,
    request: GitRepositoryRequest,
) -> Result<Vec<bitfun_core::service::git::GitWorktreeInfo>, String> {
    info!("Listing worktrees for '{}'", request.repository_path);

    GitService::list_worktrees(&request.repository_path)
        .await
        .map_err(|e| {
            error!(
                "Failed to list worktrees: path={}, error={}",
                request.repository_path, e
            );
            format!("Failed to list worktrees: {}", e)
        })
}

#[tauri::command]
pub async fn git_add_worktree(
    _state: State<'_, AppState>,
    request: GitAddWorktreeRequest,
) -> Result<bitfun_core::service::git::GitWorktreeInfo, String> {
    let create_branch = request.create_branch.unwrap_or(false);
    info!(
        "Adding worktree for branch '{}' in '{}' (create_branch: {})",
        request.branch, request.repository_path, create_branch
    );

    GitService::add_worktree(&request.repository_path, &request.branch, create_branch)
        .await
        .map_err(|e| {
            error!(
                "Failed to add worktree: path={}, branch={}, create_branch={}, error={}",
                request.repository_path, request.branch, create_branch, e
            );
            format!("Failed to add worktree: {}", e)
        })
}

#[tauri::command]
pub async fn git_remove_worktree(
    _state: State<'_, AppState>,
    request: GitRemoveWorktreeRequest,
) -> Result<GitOperationResult, String> {
    let force = request.force.unwrap_or(false);
    info!(
        "Removing worktree '{}' from '{}' (force: {})",
        request.worktree_path, request.repository_path, force
    );

    GitService::remove_worktree(&request.repository_path, &request.worktree_path, force)
        .await
        .map_err(|e| {
            error!(
                "Failed to remove worktree: path={}, worktree_path={}, force={}, error={}",
                request.repository_path, request.worktree_path, force, e
            );
            format!("Failed to remove worktree: {}", e)
        })
}

// MARK: Git Repo History

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitRepoHistory {
    pub url: String,
    pub last_used: String,
    pub local_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitRepoHistoryData {
    pub repos: Vec<GitRepoHistory>,
    pub saved_at: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveGitRepoHistoryRequest {
    pub repos: Vec<GitRepoHistory>,
}

#[tauri::command]
pub async fn save_git_repo_history(
    state: State<'_, AppState>,
    request: SaveGitRepoHistoryRequest,
) -> Result<(), String> {
    let workspace_service = &state.workspace_service;
    let persistence = workspace_service.persistence();

    let data = GitRepoHistoryData {
        repos: request.repos,
        saved_at: chrono::Utc::now().to_rfc3339(),
    };

    persistence
        .save_json("git_repo_history", &data, StorageOptions::default())
        .await
        .map_err(|e| {
            error!("Failed to save git repo history: {}", e);
            format!("Failed to save git repo history: {}", e)
        })
}

#[tauri::command]
pub async fn load_git_repo_history(
    state: State<'_, AppState>,
) -> Result<Vec<GitRepoHistory>, String> {
    let workspace_service = &state.workspace_service;
    let persistence = workspace_service.persistence();

    let data: Option<GitRepoHistoryData> = persistence
        .load_json("git_repo_history")
        .await
        .map_err(|e| {
            error!("Failed to load git repo history: {}", e);
            format!("Failed to load git repo history: {}", e)
        })?;

    match data {
        Some(data) => Ok(data.repos),
        None => Ok(Vec::new()),
    }
}
